plugins/notifications-backend/src/service/handlers.ts

```diff
@@ -136,34 +192,100 @@ export async function getNotifications(
   return notifications;
 }
 
-export function getNotificationsCount(
+export async function getNotificationsCount(
   dbClient: Knex<any, any>,
+  catalogClient: CatalogClient,
   filter: NotificationsFilter,
 ): Promise<{ count: number }> {
-  const query = dbClient('messages').count('* as CNT');
+  if (filter.messageType !== 'system' && !filter.user) {
+    throw new Error(
+      'user parameter is missing in request',
+    );
+  }
+  
+  let userGroups: string[] = [];
 
-  addFilter(query, filter);
+  if (filter.user) {
+    userGroups = await getUserGroups(catalogClient, filter.user)
+  }
 
-  const ret = query.then(count => {
-    const msgcount = count[0].CNT as number;
-    return { count: msgcount };
-  });
+  const query = createQuery(dbClient, filter, userGroups);
+  
+  const ret = query.count('* as CNT')
+    .then(count => {
+      const msgcount = count[0].CNT as number;
+      return { count: msgcount };
+    });
 
   return ret;
 }
 
-function addFilter(query: Knex.QueryBuilder, filter: NotificationsFilter) {
+function createQuery(dbClient: Knex<any, any>, filter: NotificationsFilter, userGroups: string[]) {
+  // select messages matching filter
+  const query = dbClient('messages');
+
+  // select either system messages or messages intended for the user
+  query.where(function () {
+    if (filter.messageType !== 'user') {
+      this.where('is_system', true);
+    }
+
+    if (filter.messageType !== 'system') {
+      this.orWhere(function () {
+        this.where('is_system', false)
+          .andWhere(function () {
+            this.whereIn('id', function () {
+              this.select('message_id').from('users').where('user', filter.user)
+            });
+
+            if (Array.isArray(userGroups) && userGroups.length > 0) {
+              this.orWhereIn('id', function () {
+                this.select('message_id').from('groups')
+                  .whereIn('group', userGroups)
+              })
+            }
+        })
+      })
+    }
+  })
+
+  // filter by text
   if (filter.containsText) {
-    // eslint-disable-next-line func-names
-    query.where(function () {
+    query.andWhere(function () {
       this.whereILike('title', `%${filter.containsText}%`).orWhereILike(
         'message',
         `%${filter.containsText}%`,
       );
     });
   }
 
+  // filter by time
   if (filter.createdAfter) {
     query.andWhere('created', '>', filter.createdAfter);
   }
+
+  // join with users table to get messages status. E.g read/unread
+  query.leftJoin('users', 'messages.id', 'users.message_id')
+
+  return query;
 }
+
+function getUserGroups(catalogClient: CatalogClient, user: string): Promise<string[]> {
+  return catalogClient.getEntityByRef(`user:${user}`).then(userRef => {
+    if (!userRef) {
+      throw new Error(`user '${user}' does not exist`);
+    }
+    
+    if (userRef.spec && Array.isArray(userRef.spec.memberOf)) {
+      return userRef.spec.memberOf.map(value => {
```

Comment:
```suggestion
      return userRef.spec.memberOf.map(value => value?.toString() || '')
```

Reply:
if we want shorthand i think this is better:
value ? value.toString() : ''
it is a common syntax to many languages and it shows 2 execution paths

Reply:
The suggested syntax is to avoid the use of ternary operators which can get complicated very easily.
But it's codestyle.